pkg/build/builder/common.go

```
@@ -82,3 +89,107 @@ func updateBuildRevision(c client.BuildInterface, build *api.Build, sourceInfo *
+	containerName := fmt.Sprintf("openshift_%s-build_%s_%s_post-commit_%08x", strategyName, build.Name, build.Namespace, rand.Uint32())
+	return runScriptInContainer(dockerClient, build.Spec.PostCommit.RunBash, image, containerName)
+}
+
+// runScriptInContainer runs a Bash script in an ephemeral Docker container
+// started off image, and returns an error if the script cannot be run or
+// returns a non-zero exit code.
+func runScriptInContainer(client DockerClient, script, image, containerName string) error {
+	limits, err := cgroupResourceLimits()
+	if err != nil {
+		return err
+	}
+	return dockerRun(client, docker.CreateContainerOptions{
+		Name: containerName,
+		Config: &docker.Config{
+			Image:      image,
+			Entrypoint: []string{"bash", "-c"},
+			Cmd:        []string{script},
+		},
+		HostConfig: &docker.HostConfig{
+			// Limit container's resource allocation.
+			CPUShares: limits.CPUShares,
+			CPUPeriod: limits.CPUPeriod,
+			CPUQuota:  limits.CPUQuota,
+			Memory:    limits.Memory,
+		},
+	}, docker.LogsOptions{
+		// Stream logs to stdout and stderr.
+		OutputStream: os.Stdout,
+		ErrorStream:  os.Stderr,
+		Follow:       true,
+		Stdout:       true,
+		Stderr:       true,
+	})
+}
+
+// cgroupResourceLimit represents limits used to constrain resources allocated
+// by containers.
+type cgroupResourceLimit struct {
+	CPUShares int64
+	CPUPeriod int64
+	CPUQuota  int64
+	Memory    int64
+}
+
+// cgroupResourceLimits reads CPU and memory limits from /sys/fs/cgroup.
+func cgroupResourceLimits() (cgroupResourceLimit, error) {
+	r := intFromFileReader{}
+	cpuShares := r.read("/sys/fs/cgroup/cpu,cpuacct/cpu.shares")
+	cpuPeriod := r.read("/sys/fs/cgroup/cpu,cpuacct/cpu.cfs_period_us")
+	cpuQuota := r.read("/sys/fs/cgroup/cpu,cpuacct/cpu.cfs_quota_us")
+	memory := r.read("/sys/fs/cgroup/memory/memory.limit_in_bytes")
+	return cgroupResourceLimit{
+		CPUShares: cpuShares,
+		CPUPeriod: cpuPeriod,
+		CPUQuota:  cpuQuota,
+		Memory:    memory,
+	}, r.err
```

my code has both cases.  For some errors (overflow) we ignore the error and return maxint64 because that's "close enough" in terms of the intended limit.

in an underflow case, we return a value (because we can't *not* return a value, it's not a pointer so nil is not an option) and an error.  The value we return is basically undefined, the caller can't rely on it for anything because an error occurred.

I think @rhcarvalho's situation here is similar.  If you are returning an error, it doesn't matter what else you return, any caller who relies on it is breaking the contract.